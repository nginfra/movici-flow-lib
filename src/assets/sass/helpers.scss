.is-size-6-half {
  font-size: 0.85rem !important;
}

.is-opacity-0 {
  opacity: 0;
}

.is-opacity-1 {
  opacity: 1;
}

.text-ellipsis {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: pre;
}

// .x-overflow is used alongside with FixedPosition.ts as the overflow parent
//
.overflow,
.x-overflow {
  overflow: auto;
  padding-right: 0.5rem;
  margin-right: -0.5rem;
}

.overflow-hidden {
  overflow: hidden !important;
}

.overflow-hover {
  overflow: hidden;
  &:hover {
    overflow: auto;
    padding-right: 0.5rem;
    margin-right: -0.5rem;
  }
}

button {
  &.is-borderless {
    border: 0;
  }
  &.is-transparent {
    background-color: transparent;
    color: $black;
  }
  &.is-round {
    border-radius: 100%;
  }
  &.movici {
    background: $white;
    color: $primary;
    border: 2px solid $white-ter !important;
    &:hover {
      color: $primary;
      background: $primary-light;
      border-color: $grey-light;
    }
    &:active {
      color: $primary;
      border-color: $primary !important;
    }
  }
}

.icon.is-tiny {
  font-weight: 100;
  height: 0.75rem;
  width: 0.75rem;
}

[required] {
  label,
  .label {
    &::after {
      margin-left: 2px;
      content: '*';
      color: $grey;
      font-weight: bold;
    }
  }
}
