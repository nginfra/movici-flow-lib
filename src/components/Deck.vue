<template>
  <div id="mapbox-container">
    <div id="map" />
    <canvas id="deckgl-overlay" :style="backgroundColorStyle" @contextmenu.prevent />
    <div class="map-control map-control-zero" v-if="loaded">
      <slot name="control-zero" v-bind="slotProps" />
    </div>
    <div class="map-control map-control-left" v-if="loaded">
      <slot name="control-left" v-bind="slotProps" />
    </div>
    <div class="map-control map-control-right" v-if="loaded">
      <slot name="control-right" v-bind="slotProps" />
    </div>
    <div class="map-control map-control-bottom" v-if="loaded">
      <slot name="control-bottom" v-bind="slotProps" />
    </div>
    <template v-if="loaded">
      <slot name="map" :map="map" :view-state="modelValue" />
    </template>
  </div>
</template>

<script setup lang="ts">
import { Deck as DeckGL, Layer } from "@deck.gl/core";
import type { ControllerOptions } from "@deck.gl/core/controllers/controller";
import type { DeckProps, PickInfo } from "@deck.gl/core/lib/deck";
import { viewport, type BoundingBox } from "@mapbox/geo-viewport";
import { useMoviciSettings } from "@movici-flow-common/baseComposables/useMoviciSettings";
import { useSnackbar } from "@movici-flow-common/baseComposables/useSnackbar";
import type {
  CursorCallback,
  DeckEntityObject,
  DeckEvent,
  DeckEventCallback,
  DeckEventPayload,
  DeckMouseEvent,
  ViewState,
} from "@movici-flow-common/types";
import mapboxgl from "mapbox-gl";
import "mapbox-gl/dist/mapbox-gl.css";
import { computed, onBeforeUnmount, onMounted, ref, watch } from "vue";

const DEFAULT_VIEWSTATE = useMoviciSettings().settings.defaultViewState;

function getViewportFromBBOX(
  bounding_box: BoundingBox,
  dimensions: [number, number],
  ratio: number
) {
  const { center, zoom } = viewport(
      bounding_box,
      // we set the ratio 1/3 as we want the viewport to occupy 1/3 of the map screen
      dimensions.map((side) => side * ratio) as [number, number],
      undefined, // min zoom, default 0
      undefined, // max zoom, default 20
      undefined, // tileSize, default 256
      true // allow float for zoom
    ),
    [longitude, latitude] = center;

  return { longitude, latitude, zoom };
}

function getCanvasDimensions(map: mapboxgl.Map): [number, number] {
  const dimensions = map.getCanvas();
  return [dimensions.width, dimensions.height];
}

// type DeckSlots = "control-zero" | "control-left" | "control-right" | "control-bottom";

const { failMessage } = useSnackbar();

const props = withDefaults(
  defineProps<{
    modelValue?: ViewState;
    basemap?: string;
    accessToken?: string;
    controller?: ControllerOptions;
    layers?: Layer<unknown>[];
  }>(),
  {
    basemap: "mapbox://styles/mapbox/light-v10",
    accessToken: import.meta.env.VITE_MAPBOX_TOKEN,
    layers: () => [],
  }
);

const emit = defineEmits<{
  (e: "update:modelValue", val: ViewState): void;
}>();
const map = ref<mapboxgl.Map>();
const deck = ref<DeckGL>();
const eventListeners = ref<Record<DeckEvent, Map<string, DeckEventCallback>>>({
  click: new Map<string, DeckEventCallback>(),

  error: new Map<string, DeckEventCallback>(),
});

const loaded = ref(false);
const contextPickInfo = ref<PickInfo<DeckEntityObject<unknown>>>();
const getCursor = ref<CursorCallback>();

const showMap = computed(() => props.basemap.startsWith("mapbox://"));
const backgroundColorStyle = computed(() =>
  props.basemap.startsWith("color://") ? { "background-color": props.basemap.split("//")[1] } : {}
);

function on(event: DeckEvent, callbacks: Record<string, DeckEventCallback>) {
  Object.entries(callbacks).forEach(([key, callback]) => {
    eventListeners.value[event].set(key, callback);
  });
}

function invokeCallbacks(event: DeckEvent, payload: DeckEventPayload) {
  for (const cb of eventListeners.value[event].values() ?? []) {
    cb(payload);
  }
}

function setCursorCallback(cb?: CursorCallback) {
  getCursor.value = cb || (() => null);
}

function zoomToBBox(bounding_box: BoundingBox, ratio = 1 / 3) {
  try {
    if (map.value) {
      const viewstate = {
        ...props.modelValue,
        ...getViewportFromBBOX(bounding_box, getCanvasDimensions(map.value), ratio),
        transitionDuration: 300,
      };

      updateViewState(viewstate as ViewState);
    }
  } catch (error) {
    failMessage("Error when centering to BBOX");
    console.error(error);
  }
}

function updateViewState(viewState: ViewState) {
  deck.value?.setProps({ viewState });
  map.value?.jumpTo({
    center: [viewState.longitude, viewState.latitude],
    zoom: viewState.zoom,
    bearing: viewState.bearing,
    pitch: viewState.pitch,
  });

  emit("update:modelValue", viewState);
}

function resetContextPickInfo() {
  contextPickInfo.value = undefined;
}

const slotProps = computed(() => ({
  map: map.value,
  contextPickInfo: contextPickInfo.value,
  on: on,
  resetContextPickInfo,
  onViewstateChange: updateViewState,
  setCursorCallback,
  zoomToBBox,
}));

watch(
  () => props.basemap,
  (basemap) => {
    map.value?.setStyle(showMap.value ? basemap : (null as unknown as string));
  }
);
watch(
  () => props.layers,
  (layers) => {
    deck.value?.setProps({ layers });
  }
);

watch(
  () => props.modelValue,
  (val) => {
    val && updateViewState(val);
  }
);

function initDeck(val: ViewState) {
  return new DeckGL({
    canvas: "deckgl-overlay",
    width: "100%",
    height: "100%",
    initialViewState: val,

    onClick: (info: PickInfo<DeckEntityObject<unknown>>, ev?: DeckMouseEvent) => {
      resetContextPickInfo();
      const payload = { pickInfo: info, ev, layer: info.layer as Layer<unknown> };
      if (ev?.leftButton) {
        invokeCallbacks("click", payload);
      } else if (ev?.rightButton) {
        contextPickInfo.value = info;
      }
      invokeCallbacks("click", payload);
    },
    onError: (error: Error, layer?: Layer<unknown>) => {
      console.error(error);
      invokeCallbacks("error", { error, layer });
    },
    getCursor: ({ isHovering, isDragging }: { isHovering: boolean; isDragging: boolean }) => {
      const cursorOverride = getCursor.value?.({ isHovering, isDragging });
      if (cursorOverride) return cursorOverride;
      if (isHovering) return "pointer";
      if (isDragging) return "grabbing";
      return "grab";
    },
    controller: props.controller ?? true,
    layers: [],
    onViewStateChange: ({ viewState }: { viewState: ViewState }) => {
      updateViewState(viewState);
    },
  } as unknown as DeckProps);
}

function initMapBox(viewState: ViewState) {
  return new mapboxgl.Map({
    center: [viewState.longitude, viewState.latitude],
    zoom: viewState.zoom,
    bearing: viewState.bearing,
    pitch: viewState.pitch,
    container: "map",
    accessToken: props.accessToken,
    maxPitch: 65,
    style: props.basemap,
    attributionControl: false,
    interactive: false,
  });
}
onMounted(() => {
  map.value = initMapBox(props.modelValue || DEFAULT_VIEWSTATE);
  map.value.on("load", () => {
    map.value?.resize();
    deck.value = initDeck(props.modelValue || DEFAULT_VIEWSTATE);
    loaded.value = true;
  });
});
onBeforeUnmount(() => {
  // todo: destroy children before removing map
  // this.$children.forEach((child) => child.$destroy());
  map.value?.remove();
  deck.value?.canvas.remove();
});

defineExpose({ zoomToBBox });
</script>

<style scoped lang="scss">
#mapbox-container {
  width: 100%;
  & > #map,
  & > #deckgl-overlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100% !important;
    height: 100% !important;
  }

  .map-control-zero {
    position: absolute;
    top: 0;
    left: 0;
  }

  :deep(.map-control-left) {
    position: absolute;
    top: 10px;
    left: 10px;
    display: flex;
    flex-direction: column;
    align-content: flex-start;
    transition: transform 0.5s;
    & > * {
      margin-bottom: 8px;
      margin-right: auto;
      &:last-child {
        margin-bottom: 0;
      }
    }
  }

  :deep(.map-control-right) {
    display: flex;
    flex-direction: column;
    align-content: flex-end;
    position: absolute;
    top: 10px;
    right: 10px;
    overflow: auto;
    padding: 4px 8px 24px;
    margin-right: -8px;
    margin-left: -8px;
    max-height: calc(100% - 20px - 140px);
    & > * {
      width: 100%;
      margin-bottom: 8px;
      margin-left: auto;
      &:last-child {
        margin-bottom: 0;
      }
    }
  }

  :deep(.map-control-bottom) {
    $left-menu-size: 300px;
    position: fixed;
    bottom: 24px;
    z-index: 1;
    left: 0;
    right: 0;
    width: calc(80vw - #{$left-menu-size} - #{$menu-item-size});
    display: flex;
    justify-content: center;
    transition: transform 0.5s, width 0.5s;
  }
  .deck-tooltip {
    width: 30px;
    height: 20px;
  }

  :deep(.mapboxgl-ctrl-bottom-left) {
    right: 10px;
    left: unset;
  }
  :deep(.mapboxgl-popup) {
    z-index: 99;
    .mapboxgl-popup-content {
      padding: 18px 24px 12px 12px;
    }
    .mapboxgl-popup-close-button {
      font-family: "Font Awesome 5 Pro";
      font-size: 0px;
      content: "";
      width: 24px;
      height: 24px;
      &::before {
        font-size: 12px;
        content: "\f00d";
      }
    }
  }
  .map-control:empty {
    display: none;
  }
}
</style>
