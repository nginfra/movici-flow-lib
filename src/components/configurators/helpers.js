import { hexToColorTriple, MoviciColors } from '@/visualizers/maps/colorMaps';
export function getLegendPlaceholders(items, type, maxValue) {
    if (!items.length)
        return [];
    return items.map((_, index) => getSingleLegendPlaceholder(items, type, maxValue, index));
}
export function getSingleLegendPlaceholder(items, type, maxValue, index) {
    var _a, _b;
    if (index > items.length - 1) {
        throw RangeError('index out of bounds');
    }
    let rv = `${items[index]}`;
    if (type === 'range') {
        rv += ' - ' + ((_b = (_a = items[index + 1]) !== null && _a !== void 0 ? _a : maxValue) !== null && _b !== void 0 ? _b : '');
    }
    return rv;
}
export function recalculateColorMapping(params) {
    const colors = recalculateColors(params);
    return recalculateMappingValues(params).map((val, idx) => [val, colors[idx]]);
}
function recalculateColors(params) {
    const rv = Array(params.nSteps).fill(hexToColorTriple(MoviciColors.WHITE));
    params.colors.forEach((c, idx) => (rv[idx] = c));
    return rv;
}
export function recalculateMappingValues(params) {
    var _a;
    if (!params.values.length && (params.minValue === undefined || params.maxValue === undefined)) {
        throw new Error('Insufficient input to recalculate values');
    }
    if (params.values.length === params.nSteps && !params.forceRecalculateValues) {
        return params.values;
    }
    const min = (_a = params.minValue) !== null && _a !== void 0 ? _a : params.values[0];
    const rv = [];
    let max = params.values[params.values.length - 1];
    let nSteps = params.nSteps - 1;
    if (params.maxValue !== undefined) {
        max = params.maxValue;
        if (!params.maxValueAsLastValue) {
            nSteps++;
        }
    }
    const stepSize = (max - min) / nSteps;
    for (let index = 0; index < params.nSteps; index++) {
        const val = min + stepSize * index;
        rv.push(parseFloat(val.toFixed(3)));
    }
    return rv;
}
