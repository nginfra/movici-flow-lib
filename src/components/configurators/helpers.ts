import { ColorMapping, RGBAColor } from '~flow/types';
import { hexToColorTriple, MoviciColors } from '~flow/visualizers/maps/colorMaps';

export type PlaceholderType = 'single' | 'range';

export type RecalculateMappingValueParams = {
  values: number[];
  nSteps: number;
  minValue?: number;
  maxValue?: number;
  forceRecalculateValues?: boolean;
  maxValueAsLastValue?: boolean;
};

export function getLegendPlaceholders(
  items: number[],
  type: PlaceholderType,
  maxValue?: number
): string[] {
  if (!items.length) return [];
  return items.map((_, index) => getSingleLegendPlaceholder(items, type, maxValue, index));
}

export function getSingleLegendPlaceholder(
  items: number[],
  type: PlaceholderType,
  maxValue: number | undefined,
  index: number
): string {
  if (index > items.length - 1) {
    throw RangeError('index out of bounds');
  }
  let rv = `${items[index]}`;
  if (type === 'range') {
    rv += ' - ' + (items[index + 1] ?? maxValue ?? '');
  }
  return rv;
}

export function recalculateColorMapping(
  params: RecalculateMappingValueParams & { colors: RGBAColor[] }
): ColorMapping {
  const colors = recalculateColors(params);
  return recalculateMappingValues(params).map((val, idx) => [val, colors[idx]]);
}

function recalculateColors(params: { colors: RGBAColor[]; nSteps: number }): RGBAColor[] {
  const rv = Array(params.nSteps).fill(hexToColorTriple(MoviciColors.WHITE));
  params.colors.forEach((c, idx) => (rv[idx] = c));
  return rv;
}

export function recalculateMappingValues(params: RecalculateMappingValueParams): number[] {
  if (!params.values.length && (params.minValue === undefined || params.maxValue === undefined)) {
    throw new Error('Insufficient input to recalculate values');
  }

  if (params.values.length === params.nSteps && !params.forceRecalculateValues) {
    return params.values;
  }

  const min = params.minValue ?? params.values[0];
  const rv = [];
  let max = params.values[params.values.length - 1];
  let nSteps = params.nSteps - 1;

  if (params.maxValue !== undefined) {
    max = params.maxValue;
    if (!params.maxValueAsLastValue) {
      nSteps++;
    }
  }
  const stepSize = (max - min) / nSteps;
  for (let index = 0; index < params.nSteps; index++) {
    const val = min + stepSize * index;
    rv.push(parseFloat(val.toFixed(3)));
  }
  return rv;
}
