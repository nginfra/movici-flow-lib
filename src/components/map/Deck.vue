<template>
  <div>
    <div id="mapbox-container">
      <div id="map" />
      <canvas id="deckgl-overlay" />
      <div class="map-control-zero" v-if="loaded">
        <slot name="control-zero" :map="map" :on-viewstate-change="updateViewState" />
      </div>
      <div class="map-control-left" v-if="loaded">
        <slot name="control-left" :map="map" :on-viewstate-change="updateViewState" />
      </div>
      <div class="map-control-right" v-if="loaded">
        <slot name="control-right" :map="map" :on-viewstate-change="updateViewState" />
      </div>
      <div class="map-control-bottom" v-if="loaded">
        <slot name="control-bottom" :map="map" :on-viewstate-change="updateViewState" />
      </div>
      <template v-if="loaded">
        <slot name="map" :map="map" :deck="deck" :view-state="value" />
      </template>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from 'vue-property-decorator';
import mapboxgl from 'mapbox-gl';
import 'mapbox-gl/dist/mapbox-gl.css';
import { Deck as DeckGL, Layer } from '@deck.gl/core';
import { DeckProps, PickInfo } from '@deck.gl/core/lib/deck';
import { hasOwnProperty } from '../../utils';
import { CameraOptions, Nullable } from '../../types';

const DEFAULT_VIEWSTATE = {
  latitude: 51.992381,
  longitude: 4.3649092,
  zoom: 10,
  bearing: 0,
  pitch: 0
};

@Component({
  name: 'Deck'
})
export default class Deck extends Vue {
  @Prop([Object]) readonly value!: Nullable<CameraOptions>;

  @Prop({
    type: String,
    default: 'mapbox://styles/mapbox/light-v10'
  })
  readonly basemap!: string;

  @Prop({
    type: String,
    default: process.env.VUE_APP_MAPBOX_TOKEN
  })
  readonly accessToken!: string;

  @Prop({
    type: Array,
    default() {
      return [];
    }
  })
  readonly layers!: Layer<unknown>[];
  map: mapboxgl.Map | null = null;
  deck: DeckGL | null = null;
  loaded = false;

  @Watch('basemap')
  setStyle() {
    this.map?.setStyle(this.basemap);
  }

  mounted() {
    const viewState = this.value || DEFAULT_VIEWSTATE;
    this.map = new mapboxgl.Map({
      center: [viewState.longitude, viewState.latitude],
      zoom: viewState.zoom,
      bearing: viewState.bearing,
      pitch: viewState.pitch,
      container: 'map',
      accessToken: this.accessToken,
      maxPitch: 65,
      style: this.basemap,
      attributionControl: false,
      interactive: false
    });

    this.map.on('load', () => {
      this.loaded = true;
    });

    this.deck = new DeckGL(({
      canvas: 'deckgl-overlay',
      width: '100%',
      height: '100%',
      initialViewState: viewState,
      // @ts-expect-error
      getCursor: ({ isHovering }) => {
        return isHovering ? 'pointer' : 'initial';
      },
      controller: true,
      layers: [],
      onViewStateChange: ({ viewState }: { viewState: CameraOptions }) => {
        this.updateViewState(viewState);
      },
      getTooltip(info: PickInfo<unknown>) {
        const object = info.object;
        if (object && typeof object === 'object' && hasOwnProperty(object, 'onHoverText')) {
          return `${object.onHoverText}`;
        }
      }
    } as unknown) as DeckProps);
  }

  @Watch('layers')
  renderDeck() {
    if (!this.deck) {
      throw new Error('No Deck to render');
    }
    if (!this.map) {
      throw new Error('Cannot render deck when Map is not initialized');
    }

    this.deck.setProps({ layers: this.layers });
  }

  @Watch('value')
  updateViewState(viewState: CameraOptions) {
    this.deck && this.deck.setProps({ viewState });
    this.map?.jumpTo({
      center: [viewState.longitude, viewState.latitude],
      zoom: viewState.zoom,
      bearing: viewState.bearing,
      pitch: viewState.pitch
    });
    this.$emit('input', viewState);
  }

  beforeDestroy() {
    // this was triggering errors
    // need to destroy children, so that layer removal is done before the map itself is removed
    this.$children.forEach(child => child.$destroy());
    this.map?.remove();
    // this.deck?.canvas.remove();
  }
}
</script>

<style scoped lang="scss">
#mapbox-container {
  width: 100%;
  height: 100vh;

  & > #map,
  & > #deckgl-overlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100% !important;
    height: 100% !important;
  }

  .map-control-zero {
    position: absolute;
    top: 0;
    left: 0;
  }

  ::v-deep {
    .map-control-left {
      position: absolute;
      top: 10px;
      left: 10px;
      transition: transform 0.5s;
      & > * {
        margin-bottom: 8px;
        margin-right: auto;
        &:last-child {
          margin-bottom: 0;
        }
      }
    }

    .map-control-right {
      position: absolute;
      top: 10px;
      right: 10px;
      & > * {
        margin-bottom: 8px;
        margin-left: auto;
        &:last-child {
          margin-bottom: 0;
        }
      }
    }

    .map-control-bottom {
      position: absolute;
      bottom: 24px;
      height: 100px;
      z-index: 1;
      left: 0;
      right: 0;
      width: calc(100vw - #{$left-menu-size} - #{$menu-item-size});
      $left-menu-size: 300px;
      display: flex;
      justify-content: center;
      transition: transform 0.5s, width 0.5s;
    }
  }

  .deck-tooltip {
    width: 30px;
    height: 20px;
  }

  ::v-deep {
    .mapboxgl-ctrl-bottom-left {
      right: 10px;
      left: unset;
    }
    .mapboxgl-popup {
      z-index: 99;
      .mapboxgl-popup-content {
        padding: 18px 24px 12px 12px;
      }
      .mapboxgl-popup-close-button {
        font-family: 'Font Awesome 5 Pro';
        font-size: 0px;
        content: '';
        width: 24px;
        height: 24px;
        &::before {
          font-size: 12px;
          content: '\f00d';
        }
      }
    }
  }
}
</style>
