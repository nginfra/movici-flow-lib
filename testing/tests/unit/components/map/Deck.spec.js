jest.mock('@deck.gl/core');

import Deck from '@movici-flow-common/components/map/Deck';
import { Deck as DeckGL } from '@deck.gl/core';
import { createComponentWrapper } from '../../helpers';

describe('Deck.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = createComponentWrapper(Deck);
  });
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  it('creates a Deck.gl instance', () => {
    expect(wrapper.vm.deck).toBeInstanceOf(DeckGL);
  });
});
