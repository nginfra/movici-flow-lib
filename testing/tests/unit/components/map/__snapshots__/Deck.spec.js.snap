// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck.vue renders correctly 1`] = `
<div>
  <div
    id="mapbox-container"
  >
    <div
      id="map"
    />
     
    <canvas
      id="deckgl-overlay"
    />
     
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
  </div>
</div>
`;
